File: lib/data/repositories/registered_library_repository_impl.dart
import 'dart:convert';

import 'package:libcheck/domain/models/library.dart';
import 'package:libcheck/domain/repositories/local_storage_repository.dart';
import 'package:libcheck/domain/repositories/registered_library_repository.dart';

class RegisteredLibraryRepositoryImpl implements RegisteredLibraryRepository {
  static const _storageKey = 'registered_libraries';

  final LocalStorageRepository _localStorage;

  RegisteredLibraryRepositoryImpl(this._localStorage);

  @override
  Future<List<Library>> getAll() async {
    final jsonString = await _localStorage.getString(_storageKey);
    if (jsonString == null) return [];

    try {
      final jsonList = jsonDecode(jsonString) as List<dynamic>;
      return jsonList
          .map((e) => Library.fromJson(e as Map<String, dynamic>))
          .toList();
    } on FormatException {
      return [];
    } on TypeError {
      return [];
    }
  }

  @override
  Future<void> saveAll(List<Library> libraries) async {
    final jsonList = libraries.map((e) => e.toJson()).toList();
    await _localStorage.setString(_storageKey, jsonEncode(jsonList));
  }

  @override
  Future<List<Library>> add(Library library) async {
    final current = await getAll();
    if (current.contains(library)) return current;
    current.add(library);
    await saveAll(current);
    return current;
  }

  @override
  Future<List<Library>> addAll(List<Library> libraries) async {
    final current = await getAll();
    for (final library in libraries) {
      if (!current.contains(library)) {
        current.add(library);
      }
    }
    await saveAll(current);
    return current;
  }

  @override
  Future<List<Library>> remove(Library library) async {
    final current = await getAll();
    current.removeWhere((e) => e == library);
    await saveAll(current);
    return current;
  }
}

File: src/data/repositories/registeredLibraryRepositoryImpl.ts
import {
  type Library,
  librariesEqual,
  libraryFromJson,
  libraryToJson,
} from '@/domain/models/library';
import type { LocalStorageRepository } from '@/domain/repositories/localStorageRepository';
import type { RegisteredLibraryRepository } from '@/domain/repositories/registeredLibraryRepository';

const STORAGE_KEY = 'registered_libraries';

export class RegisteredLibraryRepositoryImpl
  implements RegisteredLibraryRepository
{
  private readonly localStorage: LocalStorageRepository;

  constructor(localStorage: LocalStorageRepository) {
    this.localStorage = localStorage;
  }

  async getAll(): Promise<Library[]> {
    const jsonString = await this.localStorage.getString(STORAGE_KEY);
    if (jsonString === null) return [];

    try {
      const parsed: unknown = JSON.parse(jsonString);
      if (!Array.isArray(parsed)) return [];
      return parsed.map((e) =>
        libraryFromJson(e as Record<string, unknown>),
      );
    } catch {
      return [];
    }
  }

  async saveAll(libraries: Library[]): Promise<void> {
    const jsonList = libraries.map((e) => libraryToJson(e));
    await this.localStorage.setString(STORAGE_KEY, JSON.stringify(jsonList));
  }

  async add(library: Library): Promise<Library[]> {
    const current = await this.getAll();
    if (current.some((e) => librariesEqual(e, library))) return current;
    current.push(library);
    await this.saveAll(current);
    return current;
  }

  async addAll(libraries: Library[]): Promise<Library[]> {
    const current = await this.getAll();
    for (const library of libraries) {
      if (!current.some((e) => librariesEqual(e, library))) {
        current.push(library);
      }
    }
    await this.saveAll(current);
    return current;
  }

  async remove(library: Library): Promise<Library[]> {
    const current = await this.getAll();
    const filtered = current.filter((e) => !librariesEqual(e, library));
    await this.saveAll(filtered);
    return filtered;
  }
}
